epsagon/wrappers/fastapi.py

```diff
@@ -0,0 +1,77 @@
+"""
+Tracing route for Python fastapi.
+"""
+import time
+import json
+import json.decoder
+from typing import Callable
+
+import warnings
+from fastapi.routing import APIRoute
+from fastapi import Request, Response
+
+import epsagon.trace
+from epsagon.runners.fastapi import FastapiRunner
+from epsagon.common import EpsagonWarning
+from epsagon.utils import (
+    collect_container_metadata,
+    get_traceback_data_from_exception
+)
+from ..http_filters import ignore_request
+
+
+class TracingAPIRoute(APIRoute):
+    """
+    Custom tracing route - traces each route request & response
+    """
+    def get_route_handler(self) -> Callable:
+        """
+        Gets a tracing route handler - create a runner with event details,
+        including request & response data.
+        """
+        original_route_handler = super().get_route_handler()
+        async def custom_route_handler(request: Request) -> Response:
+            """
+            Traces given request and its response.
+            :param request: to trace
+            """
+            epsagon.trace.trace_factory.switch_to_async_tracer()
+
+            if ignore_request('', request.url.path.lower()):
+                return await original_route_handler(request)
+            trace = epsagon.trace.trace_factory.get_or_create_trace()
+            trace.prepare()
+            runner = None
+            response = None
+            try:
+                body = await request.json()
+            except json.decoder.JSONDecodeError:
+                body = ''
+            try:
+                runner = FastapiRunner(time.time(), request, json.dumps(body))
+                trace.set_runner(runner)
```

Comment:
Is it possible that more than one instance does set_runner on the same active trace, overriding each other?

Reply:
no, since the callbacks will be called in different jobs and our trace module will know to create a different trace for each job 
See: `epsagon.trace.trace_factory.switch_to_async_tracer()`